Add function contract type checks/wraps.

dbp opened the pull request:
This makes our `_type` function smarter (and actually moves the bulk into a helper, `__type`), so that it can check function arguments as well as plain value arguments. This'll be key with `bigBang`, but is already useful with `animate`.

I did this in a backwards compatible way, so this should be able to be merged pretty easily (and then be used in `bigBang`, @emhoracek).

The errors, however, can be a little confusing (in the function case), and perhaps it makes sense, at this point, to replace this with a more robust contract solution? 

The place where it gets confusing is then the contract check occurs because of an error within the definition of the function. For example:

Given we have this code:

```
var foo = _type([tArrow([tNumber], tNumber)], tNumber, 
                        "foo(): Usage: foo(function(n) { ... return m; })", function(f) {   
  f("blah"); 
});
```

Now this is clearly wrong, as we said the argument was supposed to be a number to number function, but we're calling it with the string. The problem with the current implementation is that the only error we even give is the usage message. So when you call:

```
foo(function(n) { return n; })
```

Which is perfectly valid (according to the type), you get the error. This is a known problem (called "Blame"), but perhaps it's okay, at least for now? As provided we implement _our_ functions correctly, the only places where the errors will come up _will_ be due to errors by users of `tlc.js`.

dbp committed:
Add function contract type checks/wraps.

emhoracek commented:
@dbp Wow this is really cool!!

Also it's making me realize how complicated something like Haskell type system must be! And I definitely don't want to attempt to reproduce that within tlc.js -- AND YET. :)

In bigBang, we have a "world", and in a game that's likely to be an object, but there's no reason right now that it can't just be a single number. So to express that it can be either, I'm writing tAny. But then it type checks even if the initial world is a number and the other functions take an object (or vice-versa)! So is the answer to require that world state is an object, even if it only has a single field? Or do we accept that we can't catch these types of errors, since the contract also can't catch if tObject is an object like { size: 50 } in one function and { largeness: "50" } in another. It'll still catch if a function returns "undefined" (right?), which seems to me the most likely error...

> perhaps it's okay, at least for now? 

Sure!

dbp commented:
Yeah, getting it to check "a -> (a -> image) -> (a -> key -> a) -> nothing"
would be pretty tricky. I can think of ideas for that function in
particular (because once we have a world, we can compare other things, at
least roughly, against it), but I'm not sure in general.

But, I think if we make tAny not match undefined (it does now), and add a
tUndefined or tNothing (the latter only because it seems like a better
description of the return value of a function that doesn't return, which
imagine will be the only place we actually use it), it'll probably be fine,
as like you said, it'll cover the not returning case (and a bunch of other
things, like the type of the key argument in on_key).

On Sunday, February 14, 2016, Libby notifications@github.com wrote:

> @dbp https://github.com/dbp Wow this is really cool!!
> 
> Also it's making me realize how complicated something like Haskell type
> system must be! And I definitely don't want to attempt to reproduce that
> within tlc.js -- AND YET. :)
> 
> In bigBang, we have a "world", and in a game that's likely to be an
> object, but there's no reason right now that it can't just be a single
> number. So to express that it can be either, I'm writing tAny. But then it
> type checks even if the initial world is a number and the other functions
> take an object (or vice-versa)! So is the answer to require that world
> state is an object, even if it only has a single field? Or do we accept
> that we can't catch these types of errors, since the contract also can't
> catch if tObject is an object like { size: 50 } in one function and {
> largeness: "50" } in another. It'll still catch if a function returns
> "undefined" (right?), which seems to me the most likely error...
> 
> perhaps it's okay, at least for now?
> 
> Sure!
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/the-learning-collective/tlcjs/pull/22#issuecomment-183886060
> .

dbp committed:
Distinguish between tAny and tNothing, add explicit return types.

This makes tAny not match undefined, because undefined is not a
meaningful value (and, in particular, if you forget to return from a
function, an undefined is returned, so a function with return type tAny
would match functions that didn't return).

One change that this forced was to add explicit return types to all the
functions - this was intended eventually anyway, but I thought I could
avoid it for now.

dbp commented:
@emhoracek I made the change to make `tAny` _not_ match `undefined`, and added `tNothing` which _does_ match `undefined` (so now, `print` returns `tNothing`). I had to make every function have an explicit return type because of this change, but that seems positive anyway. 

If this looks good, I'll merge this. I also added a (tiny) unit test framework to the mix, as I really wanted some tests :)

emhoracek commented:
:+1: Looks good to me!
